### PR TITLE
script to generate docker container for grpc-gateway.

### PR DESCRIPTION
**Pull request by mdkess**
Example usage:

`bash generate_gateway.sh -f FunTimes.proto -s FunTimes -c mkess-test-gateway`

Then run the container with
`docker run -p 8080:80 mkess-test-gateway --backend=mkess-cool-backend-service`

### CODE REVIEW

File: all/generate_gateway.sh [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+HTTP_PORT=8080
+
+printUsage() {
+  echo "Generates a docker image of a grpc-gateway server for a given service proto."  
+  echo
+  echo "The container will listen on port ${HTTP_PORT} for HTTP traffic, and proxy requests"
+  echo "to the gRPC service. In addition, it will serve a Swagger definition of its API at"
+  echo "/swagger.json"
+  echo
+  echo "Usage: generate_gateway.sh -f <service.proto> -p service_package -s ServiceName"
```
Comment (mhamrah) [inline all/generate_gateway.sh]: -p is service_package here, put below it's port.

Comment (mhamrah) [reply all/generate_gateway.sh]: we need to support multiple proto files. this bit me. it is important that all proto files are passed to `protoc` at once. `entrypoint.sh` was recently changed to better handle this.

File: all/generate_gateway.sh [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    exit 1
+fi
+
+if [[ -z $CONTAINER ]]; then
+    echo "Error: You must specify the Docker container name"
+    printUsage
+    exit 1
+fi
+
+if [[ -z $GRPC_SERVICE_HOST ]]; then
```
Comment (mhamrah) [inline all/generate_gateway.sh]: I don't think this and GRPC_SERVICE_PORT should be set here. I think the must be configured through env variables, or via flags on the container's entrypoint. https://github.com/spf13/viper has a good configuration hierarchy, and the main.go method should ensure these are set and present.

File: all/generate_gateway.sh [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+  SignalRunner(
+    func() {
+      fmt.Printf("launching http server on %v\n", server.Addr)
+      if err := server.ListenAndServe(); err != nil {
+        log.Fatalf("Could not start http server: %v", err)
+      }
+    },
+    func() {
+      shutdown, _ := context.WithTimeout(ctx, 10*time.Second)
```
Comment (bobbytables) [inline all/generate_gateway.sh]: Is there a reason you're not getting the CancelFunc from this and defer'ing it being called?

**Comment by bobbytables** [comment]
Make sure you update the README too to change the "ideally it would include a container blah blah" at the end.

File: all/generate_gateway.sh (hunk 1 above)
Comment (mdkess) [reply all/generate_gateway.sh]: Fixed yo.

**Comment by mdkess** [comment]
I updated this - it generates the image for you. Right now it has to run on the client's machine, next step is to run it in the container (i.e. using /defs or whatever to mount the dir).

File: all/generate_gateway.sh (hunk 3 above)
Comment (mhamrah) [reply all/generate_gateway.sh]: There's no use for the cancel func in any scope, is there?

**Comment by mhamrah** [comment]
+1. my comment is more of a nit, but i think it will be beneficial.

also, we can probably forego putting this in a container?

**Comment by mhamrah** [comment]
Can this be merged?

**Pull request merged**